sql/2020/03_Markup/pages_almanac_by_device.sql

```diff
@@ -0,0 +1,111 @@
+#standardSQL
+# pages almanac metrics grouped by device
+
+# real run estimated at $4.08 and took 48 seconds
+
+# to speed things up there is only one js function per custom metric property. It returns a STRUCT with all the data needed
+# current test gathers 3 bits of incormation from the custom petric properties
+# I tried to do a single js function processing the whole payload but it was very slow (50 sec) because of parsing the full payload in js
+# this uses JSON_EXTRACT_SCALAR to first get the custom metrics json string, and only passes those into the js functions
+# Estimate about twice the speed of the original code. But should scale up far better as the custom metrics are only parsed once.
+
+# helper to create percent fields
+CREATE TEMP FUNCTION AS_PERCENT (freq FLOAT64, total FLOAT64) RETURNS FLOAT64 AS (
```

Comment:
Idea: if these functions will be useful to many queries, we can define a reusable version that lives in the `almanac` dataset.

Reply:
That was mentioned when I came up with the function. I just went ahead and used it, knowing I could easily switch if a shared function was created. Maybe next year.